### ISSUE TITLE
GitHub app should post PR comments even when triggered by a push event

### ISSUE THREAD

**Issue by aviflat** [pulumi/actions#621]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

My GitHub Actions (GHA) workflow is running `pulumi preview` on `pull_request` events, and then once a PR is merged, it’s also triggered by the `push` event to `main` — in which case it runs `pulumi up`.

In the former scenario, running in response to `pull_request` events, meaning a PR is open and is the context of the run, the App posts helpful comments with resource diffs to the PR for every changed stack — as expected.

However, post-merge, the same workflow runs again, this time triggered by `push` to `main`. And for these runs, the App doesn’t post comments back to the PR.

I think I understand why, but it would be super-helpful to open a PR and verify right there that it was successfully and correctly “deployed” (and see when, what changed, etc) without having to go elsewhere.

So please consider posting comments to the PR in this situation.

I think it should be doable, because you can normally extract the PR number from the title of the merge commit.

### Affected area/feature

GitHub App

----

Thank you!

**Comment by AaronFriel** [comment]
Hey @aviflat thanks for reporting! I think we would need to solve for finding the PR that caused the push action - do you know if this is something GitHub Actions already supports, or would you consider submitting a PR to implement this?

**Comment by aviflat** [comment]
My impression is that that info is not right at hand in that context. I have an action where I need to retrieve this myself, so I’ve been thinking about this. I think the easiest way to determine the PR would be by examining the title (first line) of the commit message of the HEAD commit. Usually it contains `#nn` e.g. `#42` being the PR number. So it shouldn’t be hard to use a regex to retrieve that and then you can retrieve more details on the PR via the GitHub API, if needed.

Does that make sense?

**Comment by aviflat** [comment]
Alternatively, and probably more reliable, GitHub’s REST API has a resource [Pull Requests associated with a commit](https://docs.github.com/en/rest/commits/commits#list-pull-requests-associated-with-a-commit). Since the SHAs of the commits that are included in the push, are available in the runtime context of the action when triggered by a `push` event, it should be fairly straightforward to retrieve the PR number, URL, etc via that resource.

**Comment by AaronFriel** [comment]
It doesn't look like GitHub workflows payloads gives us that information, and I'd prefer a precise link between the push event and exactly one PR that caused it: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push

I think we could infer this, but we need to handle:

* Merge commits, one of whose parents should be the tip of the PR's branch
* Squashed merges, whose commit hash will only appear on the target branch's history
* When the PR's commits are from another repository, and the API fails to find the related commit as a result

If we were to implement this feature, we'd want it to be as reliable as commenting on the PR's workflow. A heuristic approach that could result in the action sometimes commenting on the wrong PRs - detecting on the wrong `#nnnn` - or making no comment at all - squash commits - wouldn't meet our reliability goals.

**Comment by AaronFriel** [comment]
I've created a discussion item for GitHub Actions to implement this feature here:

- https://github.com/github/feedback/discussions/17344

**Comment by AaronFriel** [comment]
Got it, is there an issue for upsert not working? That does seem like something we should handle. Was the matrix running the same stack up multiple times, or a different stack for each job the matrix created?

**Comment by aviflat** [comment]
> Got it, is there an issue for upsert not working? That does seem like something we should handle.

I didn’t open a new issue because I wasn’t sure if maybe #563 wasn’t the same problem. I couldn’t tell from the way the issue was written, as I mentioned in [a comment I posted last week](https://github.com/pulumi/actions/issues/563#issuecomment-1136413921).

> Was the matrix running the same stack up multiple times, or a different stack for each job the matrix created?

The matrix has one dimension: `stack` — so each, uh, “instance” of the job passes a unique stack name to the action.

**Comment by AaronFriel** [comment]
Thanks, that does seem like the same issue. I'll close this issue & revisit if GitHub Actions has a feature or the service team on the GitHub App issue comes up with a solution we can share.
